odl/solvers/functional/derivatives.py

```diff
@@ -0,0 +1,326 @@
+
+        if not isinstance(functional.domain, FnBase):
+            raise TypeError('`functional.domain` has to be a `FnBase` '
+                            'instance')
+
+        self.functional = functional
+        if step is None:
+            # Use half of the number of digits as machine epsilon, this
+            # "usually" gives a good balance between precision and numerical
+            # stability.
+            self.step = np.finfo(functional.domain.dtype).eps ** (1 / 2)
+        else:
+            self.step = float(step)
+
+        self.method = str(method).lower()
+        if method not in ('backward', 'forward', 'central'):
+            raise ValueError('`method` not understood')
+
+        Operator.__init__(self, functional.domain, functional.domain,
+                          linear=functional.is_linear)
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        # The algorithm takes finite differences in one dimension at a time
+        # reusing the dx vector to improve efficiency.
+        dfdx = self.domain.zero()
+        dx = self.domain.zero()
+
+        if self.method == 'backward':
+            fx = self.functional(x)
+            for i in range(self.domain.size):
+                dx[i - 1] = 0
+                dx[i] = self.step
+                dfdx[i] = fx - self.functional(x - dx)
+        elif self.method == 'forward':
+            fx = self.functional(x)
+            for i in range(self.domain.size):
+                dx[i - 1] = 0  # reset step from last iteration
+                dx[i] = self.step
+                dfdx[i] = self.functional(x + dx) - fx
+        elif self.method == 'central':
+            for i in range(self.domain.size):
+                dx[i - 1] = 0  # reset step from last iteration
+                dx[i] = self.step / 2
+                dfdx[i] = self.functional(x + dx) - self.functional(x - dx)
+        else:
+            raise RuntimeError('unknown method')
+
+        dfdx /= self.step
+        return dfdx
+
+    def derivative(self, point):
+        """Return the derivative in point.
+
+        The derivative of the gradient is often called the hessian.
+
+        Parameters
+        ----------
+        point : `domain` `element-like`
+            The point that the derivative should be taken in.
+
+        Returns
+        -------
+        derivative : `NumericalDerivative`
+            Numerical estimate of the derivative. Uses the same method as this
+            operator does, but with half the number of significant digits in
+            the step size in order to preserve numerical stability.
+
+        Examples
+        --------
+        Compute a numerical estimate of the derivative of the L2^2 norm:
+
+        >>> import odl
+        >>> space = odl.rn(3)
+        >>> func = odl.solvers.L2NormSquared(space)
+        >>> grad = NumericalGradient(func)
+        >>> hess = grad.derivative([1, 1, 1])
+        >>> hess([1, 0, 0])
```

Comment:
This rather length writeup certainly could make do for the math documentation. Also we should write down some of the rules that we make use of (the chain rule for the frechet derivative is not obvious to everyone).

I opened an issue on this here #632.